### ISSUE TITLE
Suggested security improvements

### ISSUE THREAD

**Issue by scribblemaniac** [laurent-clouet/sheepit-client#27]
Before starting, I would like to make it clear that I am not an expert in this area, however these are a few things that I know should be improved based off of the knowledge that I do have.

My biggest concern is that currently account passwords are being stored in plaintext in the config file. I would like to see it store a token generated and stored on the server and passed back to the client for addition to the config file. Then for future login sessions, the token could be passed instead of a password. A logout button would remove the token from the config and send a request to the server to revoke it as well. This would provide the following advantages that I can think of:
1. Password is not being stored permanently on the computer and is not visible to users or malicious programs.
2. Allows the auto-login session to be revoked by the server. This opens up the possibility of setting an expiry time, usage limit, or manual revoking all from the server-side.

My second, and less important issue is with the get requests. While they are both fairly secure over https, post still has the advantage of typically not being logged (see [here](https://security.stackexchange.com/questions/59334/get-over-https-vs-post-over-https) for more details) and should be used instead.

As for the implementation of these features, I would not have brought this up if I was not willing to give a helping hand in this. I obviously cannot work on the server-side implementation of the token generation, token storage, or post request handling since I don't have access to the server-side. However, I will gladly try and help with the client-side implementation if needed. Please tell me if you decide to follow my suggestions and what I can do to help.

**Comment by laurent-clouet** [comment]
A similar system is already in place for the "web session" for the auto-login on the server. It's based on cookie so it should adapted to the client.

For the GET request, i started a patch to replace some by a POST request, (at least the auth request), it should be done in the next days.

**Comment by scribblemaniac** [comment]
Good to hear. Do the session cookies expire on the server-side however? If they do expire, then we would be unable to use it indefinitely as we do with the password right now. On the other hand, if it doesn't expire than it opens things up for replay attacks.

I'm beginning to think it would be better to store the password in Credential Manager or Keychain for Windows and Mac respectively. That still leaves Linux systems, which don't necessarily have an equivalent system in place. In my mind, the best balance of security and ux could be achieved by using these password services when possible and falling back to the web login sessions for string passwords on the machine for only a week (or however long the session tokens last on the server-side). I will look into things a bit more when I get the chance and perhaps try and implement support for components of this idea when I get a chance. In the mean time I would love to hear your thoughts on adding support for services such as Credential Manager and Keychain.

**Comment by artemisart** [comment]
On linux we could chmod the config file to 600 to prevent other user accessing the password, but that doesn't stop other programs launched in the user session from reading the password.

**Comment by scribblemaniac** [comment]
That is good idea, and it has already been implemented [here](https://github.com/laurent-clouet/sheepit-client/blob/8964e4753dbf2fea417e5cdce01d0c00e276f23d/src/com/sheepit/client/SettingsLoader.java#L134-L136).

**Comment by scribblemaniac** [comment]
Again, I still have to do some testing, but it looks like [GNOME Keyring](https://wiki.gnome.org/Projects/GnomeKeyring) may do the trick for Linux.

**Comment by Nateowami** [comment]
I think using a key ring and auto-expiring tokens is overkill. With the password in plain text it's possible to take over someone's account. But all the token needs to be allowed to do (as I understand) is receive projects and render them. What's the worst an attacker could do? Give you render credits? Or do you have another threat model I'm not thinking of?

The password, of course, should be stored only on the server, and properly hashed with e.g. brcypt (see http://plaintextoffenders.com/faq/devs).

**Comment by scribblemaniac** [comment]
@Nateowami The tokens are not the issues, they are of course necessary and you understand their purpose more or less correctly. The issue is with the local storage of the password. Anyone who has access to the user's computer can read the password. With the password you can gain complete control of the account, use credits for personal rendering, steal lock out user, etc. Tokens can do some of these things as well depending on the implementation. Admittedly most of these things are relatively benign when compared with some other authenticated services, but that is no reason to neglect security. If you do require a larger motivation, consider this: many people use the same password for multiple services. Thus a compromise of this password can potentially become a much more serious matter as someone uses that password to gain access to an email and use that email to gain access to further accounts through password recovery mechanisms. The same applies to things like workplace access, online banking, etc.

> The password, of course, should be stored only on the server, and properly hashed with e.g. brcypt (see http://plaintextoffenders.com/faq/devs).

I completely agree, and the primary issue here is that this is not the case. The password is also being stored, unencrypted on the user's computer. Here's a fun experiment. If you're on a UNIX computer, try typing this into Terminal:

    grep -e password ~/.sheepit.conf

It's that easy to steal someone's password from sheepit. Really pathetic.

As a side note in case you are unaware, I [have quit this project](https://github.com/laurent-clouet/sheepit-client/pull/29#issuecomment-133766930), so while I do have an opinion on the security of this project, I have no intention of ever doing anything about it here. I have plans for something that will solve this and many of SheepIt's other flaws, however it will probably few months before I get a chance to start working on it :wink:

**Comment by Nateowami** [comment]
@scribblemaniac Sorry, I was unclear. My meaning was that the password doesn't need to be stored locally at all, in a key ring or otherwise (and should not be stored unhashed anywhere but the user's brain). 

Rather, what I meant to say is that the submission of projects and account management can be done through the web interface, and the rendering of frames can be done with tokens. The tokens would *only* be good for adding credits *not* spending them or other account management. If tokens expired weekly people who leave servers running for weeks on end would have to keep renewing the session, which would be annoying. That would be my suggestion anyway, based on my understanding of how the system works.

Ping me if/when you start a related project. I'm pretty busy and can certainly sympathize with others who can't keep up with their projects, but I may be able to help. I think it's healthy to have multiple networks.

**Comment by scribblemaniac** [comment]
@Nateowami Ah I see what you mean now and I think that is an excellent suggestion. I definitely missed this approach when I initially wrote this issue. I suppose it could be an issue that the application can't be extended to include project management capability, but that is probably better left on the website anyway.

I'll be sure to ping you when I start my project! I'm glad to hear your interested, that alone helps with my motivation :smile:

**Comment by Nateowami** [comment]
I just noticed on the [profile page](https://www.sheepit-renderfarm.com/account.php?mode=profile) you can create a "public key" which is essentially what I was proposing. It's just not the default.

Also, if account and project management from the client were needed, you could have a separate log in  for that which would work like the web interface and not store the password (just a temporary session cookie).

**Comment by ChlorideCull** [comment]
>It's that easy to steal someone's password from sheepit. Really pathetic.

And quite a lot of other services, actually, more or less. The guys behind Pidgin [wrote a bit about this.](https://developer.pidgin.im/wiki/PlainTextPasswords) The proper way of securing it is just `chmod 600 ~/.sheepit.conf`, which is already done AFAIK, if the password is needed to be stored.

As for remote users grabbing it, if they have the ability to pull random files they can most likely already grab or place other, more lucrative things. Not gonna go into the rabbit hole of that.

----

>[the password] should not be stored unhashed anywhere but the user's brain

Hopefully it's stored in a password manager because it's randomly generated for just that site, but a man can dream.

**Comment by eyeballcode** [comment]
What you could do is not use your password. Instead, generate a client token from the website and use that instead. If someone steals your token, all they can do is render frames for you as they can't log into the website

**Comment by laurent-clouet** [comment]
I choose the solution where on login a public key is giving to the renderer and stored as a password.
It's been pushed, the 'official' release should be this week with 5.1672.3030
